### ISSUE TITLE
"normal" syntax coloring sometimes uses color left over from prompt

### ISSUE THREAD

**Issue by tbodt** [fish-shell/fish-shell#6563]
To reproduce, type the following at a prompt, then press the up arrow:
```
function fish_prompt
    set_color green
    printf 'prompt> '
end
set -l fish_color_command normal
set -l fish_color_param normal
```
Recording: https://asciinema.org/a/297139

**Comment by ridiculousfish** [comment]
Hunh, I assumed this was just shown in the default color of the terminal emulator, but it's not, it really is from the prompt. Nice find.

**Comment by ridiculousfish** [comment]
Nice find, it's real. The trouble is that we capture the prompt function's output and write it unconditionally, so we don't know what color is set. An easy fix would be to emit a reset after printing the prompt, but in most cases that would be a useless expense.

I think the right fix here is to parse the prompt output so we can know whether we need a reset. Parsing the prompt would also help us truncate it (#904).

**Comment by zanchey** [comment]
As a workaround, you can put `set_color normal` at the end of the prompt.

**Comment by IlanCosman** [comment]
> I think the right fix here is to parse the prompt output so we can know whether we need a reset.

That sounds to me as though that would be a lot more expensive than the ~45 microseconds a `set_color normal` takes on my machine. Also, we are currently passing off the `set_color normal` responsibility to `fish_prompt`, so it's being run anyway. Thus adding it to Fish and then notifying prompt authors to remove it would be a net gain of 0 microseconds.

**Comment by ridiculousfish** [comment]
Right, for local shells it doesn't matter much. Slow / laggy ssh connections are the main reason for minimizing the amount of output.

**Comment by IlanCosman** [comment]
That's a good point I hadn't considered 👍 However my second point still applies there, that if we're asking `fish_prompt` to reset the colors, then we're already taking that performance hit. Moving the responsibility up a level doesn't add any extra time. It might even save time since we'd be running C++ code instead of interpreted Fish script right?
